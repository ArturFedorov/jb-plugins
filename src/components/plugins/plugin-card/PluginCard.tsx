import React, { FunctionComponent, useMemo } from 'react';
import { NavLink } from 'react-router-dom';
import { StarRating } from '../../common/star-rating/StarRating';
import { PluginIcon } from '../plugin-icon/PluginIcon';
import styles from '../plugins.module.scss';
import { IPlugin } from '../../../shared/interfaces/models/IPlugin';
import dateFormat from '../../../shared/utils/date-util/date.util';
import { formatNumber } from '../../../shared/utils/format-util/format.util';
import pluginIcon from '../../../assets/icons/jetbrains-logo.svg';

export interface IPluginCardProps {
  plugin: IPlugin;
}

export const PluginCard: FunctionComponent<IPluginCardProps> = ({ plugin }) => {
  const formattedDate = useMemo(() => dateFormat(plugin.date), [plugin.date]);
  const memoizedStarRating = useMemo(() => <StarRating rating={plugin.rating} />, [plugin.rating]);

  return (
    <NavLink to={`/plugins/${plugin.id}`} className={styles.pluginCardLink}>
      <div className={styles.pluginCard}>
        <div className={styles.pluginCardHeader}>
          <PluginIcon iconUrl={plugin.icon || pluginIcon} className={styles.pluginCardIcon} />
          <div className={styles.pluginCardSection}>
            <h3 className="is-lighter marginless">{plugin.name}</h3>
            {memoizedStarRating}
          </div>
        </div>
        <div className={styles.pluginCardDescription}>
          <span>{plugin.description}</span>
        </div>
        <div className={styles.pluginCardFooter}>
          <span className="is-caption">{formatNumber(plugin.downloads)} downloads</span>
          <span className="is-caption">{formattedDate}</span>
        </div>
      </div>
    </NavLink>
  );
};
