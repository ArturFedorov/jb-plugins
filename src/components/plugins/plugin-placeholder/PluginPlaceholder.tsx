import React, { FunctionComponent } from 'react';
import styles from '../plugins.module.scss';

export const PluginPlaceholder: FunctionComponent = () => {
  return (
    <div className={styles.pluginCardLink}>
      <div className="ph-item">
        <div className="ph-col-12">
          <div className="ph-picture"></div>
          <div className="ph-row">
            <div className="ph-col-4"></div>
            <div className="ph-col-8 empty"></div>
            <div className="ph-col-12"></div>
          </div>
        </div>
        <div className="ph-col-2">
          <div className="ph-avatar"></div>
        </div>
        <div>
          <div className="ph-row">
            <div className="ph-col-12"></div>
            <div className="ph-col-2"></div>
            <div className="ph-col-10 empty"></div>
            <div className="ph-col-8 big"></div>
            <div className="ph-col-4 big empty"></div>
          </div>
        </div>
      </div>
    </div>
  );
};
