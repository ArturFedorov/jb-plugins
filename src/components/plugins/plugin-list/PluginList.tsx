import React, { FunctionComponent } from 'react';
import classNames from 'classnames';
import { IPlugin } from '../../../shared/interfaces/models/IPlugin';
import { PluginCard } from '../plugin-card/PluginCard';
import { PluginPlaceholder } from '../plugin-placeholder/PluginPlaceholder';
import styles from '../plugins.module.scss';
import { createArrayFromNumber } from '../../../shared/utils/format-util/format.util';

export interface IPluginListProps {
  header?: string;
  plugins: IPlugin[];
  isCentered?: boolean;
  isLoading?: boolean;
}

export const PluginList: FunctionComponent<IPluginListProps> = ({
  plugins,
  isCentered = false,
  isLoading = false,
  header = 'Most Popular'
}) => {
  return (
    <div className={styles.pluginList}>
      <h2 className={styles.pluginListHeader}>{header}</h2>
      <div
        className={classNames(styles.pluginListContent, {
          [styles.pluginListContentCentered]: isCentered
        })}
      >
        {isLoading && createArrayFromNumber(6).map((item) => <PluginPlaceholder key={item} />)}
        {!isLoading && plugins.map((plugin) => <PluginCard key={plugin.id} plugin={plugin} />)}

        {!isLoading && plugins.length === 0 && (
          <div className={classNames(styles.pluginListEmpty, 'is-secondary-text')}>
            <span>Seems that we don’t have any proper plugins for you here</span>
            <span> ¯\_(ツ)_/¯</span>
          </div>
        )}
      </div>
    </div>
  );
};
