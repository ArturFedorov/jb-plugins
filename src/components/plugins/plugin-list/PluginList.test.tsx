import React from 'react';
import { shallow } from 'enzyme';
import { DEFAULT_NUMBER_OF_PLACEHOLDERS, IPluginListProps, PluginList } from './PluginList';
import { PluginPlaceholder } from '../plugin-placeholder/PluginPlaceholder';
import { PluginCard } from '../plugin-card/PluginCard';
import { TestUtil } from '../../../shared/utils/test.util';

describe('PluginList', () => {
  const plugins = TestUtil.mockPlugins(3);
  const baseProps: IPluginListProps = {
    plugins,
    isLoading: false
  };

  it('should match the snapshot', () => {
    const component = shallow(<PluginList {...baseProps} />);
    expect(component).toMatchSnapshot();
  });

  it('should show placeholders while loading', () => {
    const component = shallow(<PluginList {...baseProps} isLoading={true} />);
    const placeholders = component.find(PluginPlaceholder);

    expect(placeholders).toHaveLength(DEFAULT_NUMBER_OF_PLACEHOLDERS);
  });

  it('should show plugins instead of placeholders after loading', () => {
    const component = shallow(<PluginList {...baseProps} />);
    const placeholders = component.find(PluginPlaceholder);
    const pluginCards = component.find(PluginCard);

    expect(placeholders).toHaveLength(0);
    expect(pluginCards).toHaveLength(plugins.length);
  });

  it('should warn user if loading is finished and no components found', () => {
    const component = shallow(<PluginList plugins={[]} isLoading={false} />);
    const warning = component.find('.is-secondary-text');
    const placeholders = component.find(PluginPlaceholder);
    const pluginCards = component.find(PluginCard);

    expect(warning.text()).toContain('Seems that we donâ€™t have any proper plugins for you here');
    expect(placeholders).toHaveLength(0);
    expect(pluginCards).toHaveLength(0);
  });
});
