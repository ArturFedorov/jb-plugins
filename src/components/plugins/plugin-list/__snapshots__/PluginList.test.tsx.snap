// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginList should match the snapshot 1`] = `ShallowWrapper {}`;
